src/Lean/LocalContext.lean

```diff
@@ -115,22 +140,47 @@ Is the local declaration an implementation-detail hypothesis
 def isImplementationDetail (d : LocalDecl) : Bool :=
   d.kind != .default
 
-def value? : LocalDecl → Option Expr
-  | cdecl ..              => none
-  | ldecl (value := v) .. => some v
+/--
+Returns the value of the `ldecl`,
+but if the `ldecl` is nondependent and `allowNondep` is false, returns `none`.
+-/
+def value? : LocalDecl → (allowNondep : Bool := false) → Option Expr
+  | ldecl (nondep := false) (value := v) .., _    => some v
+  | ldecl (nondep := true)  (value := v) .., true => some v
+  | _,                                       _    => none
 
-def value : LocalDecl → Expr
-  | cdecl ..              => panic! "let declaration expected"
-  | ldecl (value := v) .. => v
+/--
+Returns the value of the `ldecl`,
+but if the `ldecl` is nondependent and `allowNondep` is false, panics.
+-/
+def value : LocalDecl → (allowNondep : Bool := false) → Expr
+  | cdecl ..,                                _     => panic! "let declaration expected"
+  | ldecl (nondep := false) (value := v) .., _     => v
+  | ldecl (nondep := true)  (value := v) .., true  => v
+  | ldecl (nondep := true) ..,               false => panic! "dependent let declaration expected"
 
-def hasValue : LocalDecl → Bool
-  | cdecl .. => false
-  | ldecl .. => true
+/--
+Returns `true` if `LocalDecl.value?` is not `none`.
+-/
+def hasValue : LocalDecl → (allowNondep : Bool := false) → Bool
+  | cdecl ..,                    _           => false
+  | ldecl (nondep := nondep) .., allowNondep => !nondep || allowNondep
 
+/-- Sets the value of an `ldecl`, otherwise returns `cdecl`s unchanged. -/
 def setValue : LocalDecl → Expr → LocalDecl
   | ldecl idx id n t _ nd k, v => ldecl idx id n t v nd k
   | d, _                       => d
 
+/-- Sets the `nondep` flag of an `ldecl`, otherwise returns `cdecl`s unchanged. -/
+def setNondep : LocalDecl → Bool → LocalDecl
```

Comment:
@kmill Should we add a note saying that it is the caller responsibility to ensure `dep => nondep` transitions are correct?

Reply:
Added some guidelines to the docstring